## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [5]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,item_id,category,subcategory,unit_price,unit_cost,unit_weight,qty,line_number
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,EUD29711-63-6U,decor,wall art,1096.80,762.71,7.50,105,1
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,XDA18116-89-4A,consumables,pens,24.95,15.09,1.49,21,0
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,ABE59463-05-7E,software,project management,13.52,7.86,1.68,39,6
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,ZMM00836-65-0C,consumables,pens,36.42,24.85,1.23,29,3
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,BZX55559-12-3X,consumables,misc,195.10,108.17,46.43,20,1


In [24]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal', 'shipping_price', 'total_price', 'line_cost',
       'profit'],
      dtype='object')

In [7]:
# Use the describe function to gather some basic statistics
df.describe()

,client_id,order_id,order_week,order_year,unit_price,unit_cost,unit_weight,qty,line_number
count,54639.000000,5.463900e+04,54639.000000,54639.000000,54639.000000,54639.000000,54639.000000,5.463900e+04,54639.000000
mean,54837.869416,5.470190e+06,11.359139,2022.993064,136.267207,99.446073,5.004116,5.702646e+02,2.979667
std,25487.438231,2.599807e+06,7.023499,0.082997,183.873135,133.164267,5.326599,1.879552e+04,2.436320
min,10033.000000,1.000886e+06,1.000000,2022.000000,0.010000,0.010000,0.000000,0.000000e+00,0.000000
25%,33593.000000,3.196372e+06,6.000000,2023.000000,20.800000,14.840000,1.450000,3.200000e+01,1.000000
50%,53305.000000,5.496966e+06,11.000000,2023.000000,68.310000,49.890000,3.240000,6.800000e+01,3.000000
75%,78498.000000,7.733869e+06,17.000000,2023.000000,173.160000,125.570000,6.890000,1.700000e+02,5.000000
max,99984.000000,9.998480e+06,52.000000,2023.000000,1396.230000,846.270000,46.430000,3.958244e+06,9.000000


In [8]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.head()

,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,item_id,category,subcategory,unit_price,unit_cost,unit_weight,qty,line_number
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,EUD29711-63-6U,decor,wall art,1096.80,762.71,7.50,105,1
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,XDA18116-89-4A,consumables,pens,24.95,15.09,1.49,21,0
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,ABE59463-05-7E,software,project management,13.52,7.86,1.68,39,6
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,ZMM00836-65-0C,consumables,pens,36.42,24.85,1.23,29,3
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,BZX55559-12-3X,consumables,misc,195.10,108.17,46.43,20,1


In [9]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
top_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [10]:
# For the category with the most entries, which subcategory had the most entries?
top_categories = df['category'].value_counts().idxmax()
top_subcategory = df[df['category'] == top_categories]['subcategory'].value_counts().idxmax()
top_subcategory

'bathroom supplies'

In [11]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [12]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = top_clients.index.tolist()
top_5_clients

[33615, 66037, 46820, 38378, 24741]

In [13]:
# How many total units (the qty column) did the client with the most entries order order?
client_most_entries = top_5_clients[0]
total_units_ordered = df[df['client_id'] == client_most_entries]['qty'].sum()
total_units_ordered

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [14]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()

,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,item_id,category,subcategory,unit_price,unit_cost,unit_weight,qty,line_number,subtotal
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,EUD29711-63-6U,decor,wall art,1096.80,762.71,7.50,105,1,115164.00
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,XDA18116-89-4A,consumables,pens,24.95,15.09,1.49,21,0,523.95
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,ABE59463-05-7E,software,project management,13.52,7.86,1.68,39,6,527.28
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,ZMM00836-65-0C,consumables,pens,36.42,24.85,1.23,29,3,1056.18
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,BZX55559-12-3X,consumables,misc,195.10,108.17,46.43,20,1,3902.00


In [15]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df.apply(lambda row: 7 * row['unit_weight'] * row['qty'] if row['unit_weight'] * row['qty'] > 50 else 10 * row['unit_weight'] * row['qty'], axis=1)
df.head()


,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,item_id,category,subcategory,unit_price,unit_cost,unit_weight,qty,line_number,subtotal,shipping_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,EUD29711-63-6U,decor,wall art,1096.80,762.71,7.50,105,1,115164.00,5512.50
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,XDA18116-89-4A,consumables,pens,24.95,15.09,1.49,21,0,523.95,312.90
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,ABE59463-05-7E,software,project management,13.52,7.86,1.68,39,6,527.28,458.64
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,ZMM00836-65-0C,consumables,pens,36.42,24.85,1.23,29,3,1056.18,356.70
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,BZX55559-12-3X,consumables,misc,195.10,108.17,46.43,20,1,3902.00,6500.20


In [16]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df.head()

,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,category,subcategory,unit_price,unit_cost,unit_weight,qty,line_number,subtotal,shipping_price,total_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,decor,wall art,1096.80,762.71,7.50,105,1,115164.00,5512.50,131839.076250
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,consumables,pens,24.95,15.09,1.49,21,0,523.95,312.90,914.258625
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,software,project management,13.52,7.86,1.68,39,6,527.28,458.64,1077.117600
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,consumables,pens,36.42,24.85,1.23,29,3,1056.18,356.70,1543.571400
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,consumables,misc,195.10,108.17,46.43,20,1,3902.00,6500.20,11364.403500


In [17]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']  
df.head()


,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,subcategory,unit_price,unit_cost,unit_weight,qty,line_number,subtotal,shipping_price,total_price,line_cost
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,wall art,1096.80,762.71,7.50,105,1,115164.00,5512.50,131839.076250,85597.05
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,pens,24.95,15.09,1.49,21,0,523.95,312.90,914.258625,629.79
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,project management,13.52,7.86,1.68,39,6,527.28,458.64,1077.117600,765.18
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,pens,36.42,24.85,1.23,29,3,1056.18,356.70,1543.571400,1077.35
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,misc,195.10,108.17,46.43,20,1,3902.00,6500.20,11364.403500,8663.60


In [18]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
df.head()
df.tail()


,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,unit_price,unit_cost,unit_weight,qty,line_number,subtotal,shipping_price,total_price,line_cost,profit
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,...,83.13,51.60,2.25,33,8,2743.29,519.75,3564.871200,2222.55,1342.321200
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,...,206.59,175.46,11.70,47,0,9709.73,3849.30,14813.240275,12095.92,2717.320275
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,...,65.66,57.31,4.16,475,6,31188.50,13832.00,49184.896250,41054.25,8130.646250
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,...,1.48,1.22,18.04,112,0,165.76,14143.36,15632.713600,14280.00,1352.713600
54638,Dwayne,Bell,Chief Financial Officer,606-619-7667x9508,belld@galvan.com,24007,7489403,2023-01-04,1,2023,...,3.01,1.61,2.07,1031,5,3103.31,14939.19,19711.431250,16599.10,3112.331250


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [19]:
# Check your work using the totals above
order_totals = df.groupby('order_id')['total_price'].sum()

order_receipts = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

verification = {order_id: order_totals.get(order_id, 'Order not found') == receipt_total
                for order_id, receipt_total in order_receipts.items()}
verification

{2742071: False, 2173913: False, 6128929: False}

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [20]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

top_5_client_spending = df[df['client_id'].isin(top_5_clients)].groupby('client_id')['total_price'].sum()
top_5_client_spending

client_id
24741    8.226889e+07
33615    8.377309e+06
38378    1.290655e+07
46820    9.743794e+06
66037    1.025951e+07
Name: total_price, dtype: float64

In [21]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5_clients_summary = df[df['client_id'].isin(top_5_clients)].groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()

top_5_clients_summary

,client_id,total_units_purchased,total_shipping_price,total_revenue,total_profit
0,24741,239862,5126448.37,8.226889e+07,3.657999e+07
1,33615,64313,1828984.89,8.377309e+06,2.201995e+06
2,38378,73667,3429455.40,1.290655e+07,3.271830e+06
3,46820,75768,1601448.84,9.743794e+06,2.736603e+06
4,66037,43018,1395151.85,1.025951e+07,3.255032e+06


In [22]:
# Format the data and rename the columns to names suitable for presentation.
# Define the money columns. 

money_columns = ['total_units_purchased', 'total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    return value / 1_000_000

# Apply the currency_format_millions function to only the money columns. 

top_5_clients_summary[money_columns] = top_5_clients_summary[money_columns].applymap(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
top_5_clients_summary.columns = [
    'Client ID',
    'Total Units Purchased (Millions)',
    'Total Shipping Price (Millions)',
    'Total Revenue (Millions)',
    'Total Profit (Millions)'
]

top_5_clients_summary


C:\Users\sokhna.thiaw\AppData\Local\Temp\ipykernel_21832\1935773829.py:12: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  top_5_clients_summary[money_columns] = top_5_clients_summary[money_columns].applymap(currency_format_millions)


,Client ID,Total Units Purchased (Millions),Total Shipping Price (Millions),Total Revenue (Millions),Total Profit (Millions)
0,24741,0.239862,5.126448,82.268892,36.579992
1,33615,0.064313,1.828985,8.377309,2.201995
2,38378,0.073667,3.429455,12.906551,3.271830
3,46820,0.075768,1.601449,9.743794,2.736603
4,66037,0.043018,1.395152,10.259515,3.255032


In [23]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_top_5_clients_summary = top_5_clients_summary.sort_values(by='Total Profit (Millions)', ascending=False)
sorted_top_5_clients_summary

,Client ID,Total Units Purchased (Millions),Total Shipping Price (Millions),Total Revenue (Millions),Total Profit (Millions)
0,24741,0.239862,5.126448,82.268892,36.579992
2,38378,0.073667,3.429455,12.906551,3.271830
4,66037,0.043018,1.395152,10.259515,3.255032
3,46820,0.075768,1.601449,9.743794,2.736603
1,33615,0.064313,1.828985,8.377309,2.201995


In [25]:
top_5_clients_summary.to_csv('top_5_clients_summary.csv', index=False)
sorted_top_5_clients_summary.to_csv('sorted_top_5_clients_summary.csv', index=False)